libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -24,666 +24,849 @@
-        _protocol = FrSkyDPORT;
-        _initialised_uart = true;   // SerialManager initialises uart for us
-    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FRSky_SPort, 0))) {
-        // check for FRSky_SPort
-        _protocol = FrSkySPORT;
-        _gps_call = 0;
-        _fas_call = 0;
-        _vario_call = 0 ;
-        _various_call = 0 ;
-        _gps_data_ready = false;
-        _battery_data_ready = false;
-        _baro_data_ready = false;
-        _mode_data_ready = false;
-        _sats_data_ready = false;
-        _sport_status = 0;
-        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::sport_tick, void));
+/*
+ * init - perform required initialisation
+ * for Copter
+ */
+void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, const char *firmware_str, const char *frame_config_str, const uint8_t mav_type, AP_Float *fs_batt_voltage, AP_Float *fs_batt_mah, uint8_t *control_mode, uint32_t *ap_value, uint32_t *control_sensors_present, uint32_t *control_sensors_enabled, uint32_t *control_sensors_health, int32_t *home_distance, int32_t *home_bearing)
+{
+    // check for protocol configured for a serial port - only the first serial port with one of these protocols will then run (cannot have FrSky on multiple serial ports)
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_D, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_D; // FrSky D protocol (D-receivers)
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_SPort; // FrSky SPort protocol (X-receivers)
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough; // FrSky SPort and SPort Passthrough (OpenTX) protocols (X-receivers)
+        // add firmware and frame info to message queue
+        char str[50];
+		sprintf(str, "%s %s", firmware_str, frame_config_str);
+        queue_message(MAV_SEVERITY_INFO, str);
+		// save main parameters locally
+		_params.mav_type = mav_type; // frame type (see MAV_TYPE in Mavlink definition file common.h)
+		_params.fs_batt_voltage = fs_batt_voltage; // failsafe battery voltage in volts
+		_params.fs_batt_mah = fs_batt_mah; // failsafe reserve capacity in mAh
+        _ap.control_mode = control_mode; // flight mode
+        _ap.value = ap_value; // ap bit-field
+        _ap.control_sensors_present = control_sensors_present;
+        _ap.control_sensors_enabled = control_sensors_enabled;
+        _ap.control_sensors_health = control_sensors_health;
+        _ap.home_distance = home_distance;
+        _ap.home_bearing = home_bearing;
     }
 
     if (_port != NULL) {
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::tick, void));
```

Comment:
All I know is that this was suggested by @tridge who said it would be called at around 1kHz: https://groups.google.com/forum/#!topic/drones-discuss/1gwvlmR8XkA